libs/common-types/src/lib.rs

```diff
@@ -130,47 +197,166 @@ where
 	Moment: From<u64> + PartialEq + PartialOrd + Saturating + Ord + Copy,
 	TrancheId: PartialEq + PartialOrd,
 {
-	type Property = PoolRole<TrancheId, Moment>;
+	type Property = Role<TrancheId, Moment>;
 	type Error = ();
 	type Ok = ();
 
 	fn exists(&self, property: Self::Property) -> bool {
 		match property {
-			PoolRole::Borrower => self.admin.contains(AdminRoles::BORROWER),
-			PoolRole::LiquidityAdmin => self.admin.contains(AdminRoles::LIQUIDITY_ADMIN),
-			PoolRole::PoolAdmin => self.admin.contains(AdminRoles::POOL_ADMIN),
-			PoolRole::PricingAdmin => self.admin.contains(AdminRoles::PRICING_ADMIN),
-			PoolRole::MemberListAdmin => self.admin.contains(AdminRoles::MEMBER_LIST_ADMIN),
-			PoolRole::RiskAdmin => self.admin.contains(AdminRoles::RISK_ADMIN),
-			PoolRole::TrancheInvestor(id, _) => self.tranche_investor.contains(id),
+			Role::PoolRole(pool_role) => match pool_role {
+				PoolRole::Borrower => self.admin.contains(AdminRoles::BORROWER),
+				PoolRole::LiquidityAdmin => self.admin.contains(AdminRoles::LIQUIDITY_ADMIN),
+				PoolRole::PoolAdmin => self.admin.contains(AdminRoles::POOL_ADMIN),
+				PoolRole::PricingAdmin => self.admin.contains(AdminRoles::PRICING_ADMIN),
+				PoolRole::MemberListAdmin => self.admin.contains(AdminRoles::MEMBER_LIST_ADMIN),
+				PoolRole::RiskAdmin => self.admin.contains(AdminRoles::RISK_ADMIN),
+				PoolRole::TrancheInvestor(id, _) => self.tranche_investor.contains(id),
+			},
+			Role::PermissionedCurrencyRole(permissioned_currency_role) => {
+				match permissioned_currency_role {
+					PermissionedCurrencyRole::Holder(_) => {
+						self.permissioned_asset_holder.contains()
+					}
+					PermissionedCurrencyRole::Manager => {
+						self.admin.contains(AdminRoles::PERMISSIONED_ASSET_MANAGER)
+					}
+					PermissionedCurrencyRole::Issuer => {
+						self.admin.contains(AdminRoles::PERMISSIONED_ASSET_ISSUER)
+					}
+				}
+			}
 		}
 	}
 
 	fn empty(&self) -> bool {
-		self.admin.is_empty() && self.tranche_investor.is_empty()
+		self.admin.is_empty()
+			&& self.tranche_investor.is_empty()
+			&& self.permissioned_asset_holder.is_empty()
 	}
 
 	fn rm(&mut self, property: Self::Property) -> Result<(), ()> {
 		match property {
-			PoolRole::Borrower => Ok(self.admin.remove(AdminRoles::BORROWER)),
-			PoolRole::LiquidityAdmin => Ok(self.admin.remove(AdminRoles::LIQUIDITY_ADMIN)),
-			PoolRole::PoolAdmin => Ok(self.admin.remove(AdminRoles::POOL_ADMIN)),
-			PoolRole::PricingAdmin => Ok(self.admin.remove(AdminRoles::PRICING_ADMIN)),
-			PoolRole::MemberListAdmin => Ok(self.admin.remove(AdminRoles::MEMBER_LIST_ADMIN)),
-			PoolRole::RiskAdmin => Ok(self.admin.remove(AdminRoles::RISK_ADMIN)),
-			PoolRole::TrancheInvestor(id, delta) => self.tranche_investor.remove(id, delta),
+			Role::PoolRole(pool_role) => match pool_role {
+				PoolRole::Borrower => Ok(self.admin.remove(AdminRoles::BORROWER)),
```

Comment:
I wonder if we could have a `TryInto` of sorts to try and convert `PoolRole` to `AdminRoles`. Having that would dry a lot of these duplicated match cases, reduce the noise, and give room for the exception cases.

Reply:
This is a bigger change I think, fine if I add a ticket for this instead?

Reply:
https://github.com/centrifuge/centrifuge-chain/issues/780